src/odemis/driver/smaract.py

```diff
@@ -436,14 +481,17 @@ def terminate(self):
     def updateMetadata(self, md):
         if model.MD_PIVOT_POS in md:
             pivot = md[model.MD_PIVOT_POS]
-            if not (isinstance(pivot, dict) and set(pivot.keys()) == {"x", "y", "z"}):
-                raise ValueError("Invalid metadata, should be a coordinate dictionary with x, y, z keys but got %s." % (pivot,))
+            if not isinstance(pivot, dict):
+                raise ValueError("Invalid metadata, should be a coordinate dictionary but got %s." % (pivot,))
+            pivot_hw = self._to_hw_axes(pivot)
+            if not set(pivot_hw.keys()) == {"x", "y", "z"}:
+                raise ValueError("Invalid metadata, should have x, y, z keys (in hardware names) but got %s." % (pivot_hw,))
```

Comment:
You could actually simplify the format string variable to % pivot and % pivot_hw on line 485 and 488.

Reply:
No and yes. For `pivot`, we cannot, because we don't know the type. We just know it's *not* a dict. If it's a tuple `1, 2, 3`, and you have the format "blabla %s" % pivot, it will try to map `1` to the `%s` and then fail because there is nothing for `2` and `3`! That's actually why I have the habit of *always* passing a tuple, even when there is only one argument.
For `pivot_hw`, as we know it's a dict, it's technically not necessary, indeed. But I prefer to *always* pass a tuple.

The conclusion is: use f-strings ;-)